test/metabase/driver/sql_jdbc/connection_test.clj

```diff
@@ -75,50 +75,64 @@
         ;; ensure that, for any sql-jdbc drier anyway, we found *some* DB name to use in this String
         (is (not= db-nm "null"))))))
 
+(deftest same-connection-details-result-in-equal-specs-test
+  (testing "Two JDBC specs created with the same details must be considered equal for the connection pool cache to work correctly"
+    ;; this is only really a concern for drivers like Spark SQL that create custom DataSources instead of plain details
+    ;; maps -- those DataSources need to be considered equal based on the connection string/properties
+    (mt/test-drivers (sql-jdbc.tu/sql-jdbc-drivers)
+      (let [details (:details (mt/db))
+            spec-1  (sql-jdbc.conn/connection-details->spec mt/*driver* details)
+            spec-2  (sql-jdbc.conn/connection-details->spec mt/*driver* details)]
+        (is (= spec-1 spec-2))))))
+
+(defn- perturb-db-details [db]
+  (update db
+          :details
+          (fn [details]
+            (case driver/*driver*
+              :redshift
+              (assoc details :additional-options "defaultRowFetchSize=1000")
+
+              (cond-> details
+                ;; swap localhost and 127.0.0.1
+                (= "localhost" (:host details))
+                (assoc :host "127.0.0.1")
+
+                (= "127.0.0.1" (:host details))
+                (assoc :host "localhost")
+
+                :else
+                (assoc :new-config "something"))))))
```

Comment:
Moved this out of the test below to make it a little more readable. No changes to logic.